gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -178,8 +193,12 @@ private Object doResumableUpload() throws IOException {
       long writeOffset = 0;
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
+      // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+      // rewound and re-sent upon non-transient errors.
+      TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
```

Comment:
Are there any guarantees that `getCommittedWriteSize` will always return one of the previously sent offsets? I.e. is it always aligned to 2 MiB (chunk size)?

Reply:
That's not assumed in the implementation, gRPC write updates with offsets like 2, 4, 6, 8Mibs and the size returned from server can be 7Mib, in that case we resume uploading from offset 6Mibs again and Michael confirmed that GCS is smart enough to start with 7 and drop the data from 6-7.